pkg/logentry/stages/template.go

```diff
@@ -113,7 +115,25 @@ func (o *templateStage) Process(labels model.LabelSet, extracted map[string]inte
 		}
 
 	} else {
-		td := templateData{}
+		td := make(map[string]interface{})
+		for _, node := range o.template.Root.Nodes {
+			if node.Type() == parse.NodeAction {
+				// node.String() outputs value as {{ .Value }} or {{ .Value | ToUpper }}
+				// regex is used to fetch only .Value and trim is used to remove the period(.)
+				source := strings.Trim(regexp.MustCompile(`\.\S+[^\s}]`).FindString(node.String()), ".")
```

Comment:
is the function applied `| ToUpper` by who and how ?

Reply:
I mean to say if the defined template is say `warning for app {{ .Value | ToUpper }}`,  node.String() will return `{{ .Value | ToUpper }}`

Reply:
So I ran the code you could merge the two branches. And also drop the template parsing, just create a td := make(map[string]interface{}) and fill in with `.Value` and the rest of all existing keys. (I think you still need the string conversion).

Also please add more test case, like if a key is missing.

Reply:
If we drop the template parsing, we should iterate all keys of `extracted` and do `getString()` and assign it to `td`. Would this be fine? Parsing helps us to use only the values defined in template from the `extracted`

Reply:
Yes it's fine but also get rid of the if branch we should include `.Value` key is source is not empty and have the same logic for everything.

| Parsing helps us to use only the keys defined in template. So, we take only those keys from the extracted

write a test with missing keys, and you'll see it is not required.

Reply:
Didn't get you. If Source is not empty, the logic is already handled by https://github.com/grafana/loki/blob/master/pkg/logentry/stages/template.go#L90.

If we have `{{.level}} with {{.exception}}` as `template`, then how do we know if the `level` and `exception` exists in `extracted` or not without parsing?

If `level` exists in `extracted` and `exception` doesn't exist in `extracted` then should we add `exception` key to `extracted` with empty value? In this case template value becomes `$somelevel with <no value>`

Reply:
I've just pushed a commit, I let you have a look at it. Now I think you should add more test where values are missing.

I just wanted to show you a less complex way of doing this.

Thank for being patient.

Reply:
Thank you @cyriltovena. Went through your code. It's simple and clean. Thank you so much.